### ISSUE TITLE
%ifstat for 10Gig network interface is wrong

### ISSUE THREAD

**Issue by tdeneau** [sysstat/sysstat#75]
For example this shows %Ifutil to be 172% although sum of txKB/s and rxkB/s is about 298,000 kB/s while interface should be able to handle 1,300,000 kB/s.

09:47:36 AM     IFACE   rxpck/s   txpck/s    rxkB/s    txkB/s   rxcmp/s   txcmp/s  rxmcst/s   %ifutil
Average:         eth1  12302.69 200575.41    966.02 296813.87      0.00      0.00      0.00    172.44

**Comment by sysstat** [comment]
%ifutil is calculated in function compute_ifutil() in sa_common.c using interface duplex value from /sys/class/net/`<iface>`/duplex and speed value from /sys/class/net/`<iface>`/speed.

What are the contents of those files (/sys/class/net/eth1/duplex and /sys/class/net/eth1/speed) for your system?
Also sending me a binary data file created by sar (saDD) where the problem appears may help me to understand and fix it.

Thanks.

**Comment by tcolgate** [comment]
I have the same problem
duplex contains "full"
and speed contains 10000
These look right, 
But I get
```
15:54:04        IFACE   rxpck/s   txpck/s    rxkB/s    txkB/s   rxcmp/s   txcmp/s  rxmcst/s   %ifutil
15:54:05         eth2 133854.55 130922.73 160821.56  76130.39      0.00      0.00      0.00     93.43
```
Which should be 1.2Gb/s, so more like 9%

shouldn't
```
return (rx * 800 / speed
```
be 8000 ? for 1KB to b?

**Comment by tcolgate** [comment]
Opened #92

**Comment by sysstat** [comment]
No, the formula is good (800 is to translate kB to B, and then compute a %).
Yet the value displayed for %ifutil is actually wrong...
Doing the calculation manually gives me the good result (around 10%).
Could you please send me a binary data file (saDD) created by sar where the problem appears? I would like to check with real values...
Thanks.

**Comment by tcolgate** [comment]
I'm just running with -n DEV 1 100

**Comment by sysstat** [comment]
I understand, but I need some real values to figure out what is happening. So just enter "sar -o datafile -n DEV 1 100" and send the datafile created to (sysstat  at orange dot fr). Of course make sure that the problem has occurred while you were creating the file.
Thanks.

**Comment by sysstat** [comment]
Thanks to some data provided by Tom (@tdeneau), I was able to find that an integer overflow happened when the interface speed was calculated, resulting in a wrong %ifutil value displayed by sar. This bug has been fixed in sysstat version 11.1.5.
@tcolgate: Please use sysstat version 11.2.0 or later and tell me if you still have the problem.
Thanks.

**Comment by pmonrealgonzalez** [comment]
Hi, Sebastien. Can you point me to the commit/s that fix this integer overflow? I have a 100 Gbps NIC and gives a ifutil of 7160.37%:

10:20:42        IFACE   rxpck/s   txpck/s    rxkB/s    txkB/s   rxcmp/s   txcmp/s  rxmcst/s   %ifutil
10:20:43         eth1  56718.00 198449.00   4807.39 10626501.55      0.00      0.00      0.00    7160.37

**Comment by sysstat** [comment]
This is annoying, I couldn't find the commit either... Did I miss something ?
I will further investigate in the coming days...

**Comment by sysstat** [comment]
In fact the problem has been fixed in sysstat-11.1.3 (not sysstat-11.1.5).
Here is the relevant commit: de9c4ff
The fix concerns the function `compute_ifutil()` in the sa_common.c file:
```
-		speed = st_net_dev->speed * 1000000;
+		speed = (unsigned long long) st_net_dev->speed * 1000000;
```
Of course, if you are using a sysstat version older than 11.1.3, then you should consider upgrading.
OTOH if you are using a recent sysstat version (11.1.3 or newer) and you still have the problem, please tell me and also send me (sysstat at orange dot fr) a binary datafile (`/var/log/sa/saDD`) where the problem appears.

**Comment by pmonrealgonzalez** [comment]
Hi, Sebastien. That's what I needed. I just patched sysstat-10.2.1 and is working perfectly. Many thanks!
